apps/emqx_authn/test/emqx_authn_mnesia_SUITE.erl

```diff
@@ -19,146 +19,215 @@
+
+t_update_user(_) ->
+    Config = config(),
+    {ok, State} = emqx_authn_mnesia:create(Config),
+
+    User = #{user_id => <<"u">>, password => <<"p">>},
+    {ok, _} = emqx_authn_mnesia:add_user(User, State),
+
+    {error, not_found} = emqx_authn_mnesia:update_user(<<"u1">>, #{password => <<"p1">>}, State),
+    {ok,
+     #{user_id := <<"u">>,
+       is_superuser := true}} = emqx_authn_mnesia:update_user(
+                                  <<"u">>,
+                                  #{password => <<"p1">>, is_superuser => true},
+                                  State),
+
+    {ok, _} = emqx_authn_mnesia:authenticate(
+                #{username => <<"u">>, password => <<"p1">>},
+                State),
+
+    {ok, #{is_superuser := true}} = emqx_authn_mnesia:lookup_user(<<"u">>, State).
+
+t_list_users(_) ->
+    Config = config(),
+    {ok, State} = emqx_authn_mnesia:create(Config),
+
+    Users = [#{user_id => <<"u1">>, password => <<"p">>},
+             #{user_id => <<"u2">>, password => <<"p">>},
+             #{user_id => <<"u3">>, password => <<"p">>}],
+
+    lists:foreach(
+      fun(U) -> {ok, _} = emqx_authn_mnesia:add_user(U, State) end,
+      Users),
+
+    {ok,
+     #{data := [#{user_id := _}, #{user_id := _}],
+       meta := #{page := 1, limit := 2, count := 3}}} = emqx_authn_mnesia:list_users(
+                                                          #{<<"page">> => 1, <<"limit">> => 2},
+                                                          State),
+    {ok,
+     #{data := [#{user_id := _}],
+       meta := #{page := 2, limit := 2, count := 3}}} = emqx_authn_mnesia:list_users(
+                                                          #{<<"page">> => 2, <<"limit">> => 2},
+                                                          State).
+
+t_import_users(_) ->
+    Config0 = config(),
+    Config = Config0#{password_hash_algorithm => #{name => sha256}},
+    {ok, State} = emqx_authn_mnesia:create(Config),
+
+    ok = emqx_authn_mnesia:import_users(
+           data_filename(<<"user-credentials.json">>),
+           State),
+
+    ok = emqx_authn_mnesia:import_users(
+           data_filename(<<"user-credentials.csv">>),
+           State),
+
+    {error, _} = emqx_authn_mnesia:import_users(
+                   <<"/file/with/unknown.extension">>,
+                   State),
+
+    {error, _} = emqx_authn_mnesia:import_users(
+                   <<"/file/with/no/extension">>,
+                   State),
+
+    {error, _} = emqx_authn_mnesia:import_users(
+                   <<"/file/that/not/exist.txt">>,
+                   State),
+
+    {error, _} = emqx_authn_mnesia:import_users(
+                   data_filename(<<"user-credentials-malformed-0.json">>),
+                   State),
+
+    {error, _} = emqx_authn_mnesia:import_users(
+                   data_filename(<<"user-credentials-malformed-1.json">>),
+                   State),
+
+    {error, _} = emqx_authn_mnesia:import_users(
```

Comment:
Suggestion/nit: maybe we can assert the expected error, to avoid the test passing for unexpected reasons?

Reply:
Yes, it's a nice catch: one of the cases actually tested a wrong case https://github.com/emqx/emqx/pull/6093/files#diff-15b244d2c564678f5cbb7ea16ec62d726ff4a1f591a3bee62307902dcb83a554R203